src/renovation_tracker/routers/listings_router.py

```diff
@@ -0,0 +1,319 @@
+from fastapi import APIRouter, HTTPException, Depends
+from typing import Annotated
+from renovation_tracker.pydantic_models.listings import (
+    Listing,
+    ListingRead,
+    ListingUpdate,
+)
+import renovation_tracker.models as models
+from renovation_tracker.database import get_db, Session
+from selenium import webdriver
+from selenium.webdriver.chrome.options import Options
+from selenium.webdriver.chrome.service import Service as ChromeService
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.webdriver.support import expected_conditions as EC
+from webdriver_manager.chrome import ChromeDriverManager
+import time
+import re
+from bs4 import BeautifulSoup
+
+
+router = APIRouter(prefix="/listings")
+db_dependency = Annotated[Session, Depends(get_db)]
+
+
+# Create listing with custom inputs
+@router.post("/", response_model=ListingRead)
+async def create_listing(listing: Listing, db: Annotated[Session, Depends(get_db)]):
+    # Create listing using user input
+    db_listing = models.Listing(**listing.dict())
+    try:
+        db.add(db_listing)
+        db.commit()
+        db.refresh(db_listing)
+        return db_listing
+    except Exception as e:
+        db.rollback()
+        raise HTTPException(status_code=500, detail=f"Error inserting listing: {e}")
+
+
+# CREATE Listing with URL also creates Photos
+@router.post("/url", response_model=ListingRead)
+async def create_url_listing(url: str, db: Annotated[Session, Depends(get_db)]):
+    # Create listing object using web scraping helper function
+    url_return = url_listing(url)
+    try:
+        db.add(url_return["listing"])
+        db.commit()
+        db.refresh(url_return["listing"])
+        for img_url in url_return["photos_list"]:
+            db_photo = models.Photos(
+                url=img_url, listing_id=url_return["listing"].listing_id
+            )
+            db.add(db_photo)
+        db.commit()
+        return url_return["listing"]
+    except Exception as e:
+        db.rollback()
+        raise HTTPException(status_code=500, detail=f"Error inserting listing: {e}")
+
+
+# READ all Listings
+@router.get("/", response_model=list[ListingRead])
+async def read_listing(db: Annotated[Session, Depends(get_db)], limit: int = 25):
```

Comment:
Use `query()` here to enforce an upper bound on limit:

```suggestion
async def read_listing(db: Annotated[Session, Depends(get_db)], limit: int = Query(default = 25, le = 100, description = "...")):
```